### ISSUE TITLE
:typing support

### ISSUE THREAD

**Issue by cpbradley** [caelan/pddlstream#32]
Hi.  As the number of arguments in my PDDL actions have grown, I've noticed solving the underlying PDDL problem has become more and more of a limiting factor.  I believe this could be mitigated substantially if my action preconditions were typed so as to reduce the number of actions considered by the planner, however this is not supported at the moment.  Could you give some insight as to why that is the case (my intuition is that typing the output of streams might be tricky for some reason)?  If you've given any thought to adding support for this feature, do you know what would need to be changed in the codebase to enable it?

**Comment by caelan** [comment]
Types are effectively just unary predicates, both in theory as well as in practice within FastDownward. Because of this, I've decided to not support them as a separate feature since they do not add expressiveness.

The set of preconditions does directly determine the number of action instances that are grounded for a finite set of objects. Look to add (static) preconditions involving several action parameters that are dependent on each other. For example, in a robotics application, a pose `?p` is typically only well-defined for a single object `?o`, which can be indicated by `(Pose ?o ?p)` and included in action/stream conditions (https://github.com/caelan/pddlstream/blob/main/examples/pybullet/pr2/stream.pddl#L16).

I've been able to plan in domains with actions that have a large number of parameters without a performance hit by following this general strategy (https://github.com/caelan/LTAMP/blob/7e0a28dca1156b1dc6d3b9459fed58607249c3ed/plan_tools/pddl/domain.pddl#L91).
